azure/services/scalesets/scalesets.go

```diff
@@ -82,89 +76,60 @@ func (s *Service) Name() string {
-			}
-			s.Scope.SetProviderID(providerID)
-			s.Scope.SetVMSSState(fetchedVMSS)
+	_, err := s.Client.Get(ctx, spec)
+	if err == nil {
+		// We can only get the existing instances if the VMSS already exists
+		scaleSetSpec.VMSSInstances, err = s.Client.ListInstances(ctx, spec.ResourceGroupName(), spec.ResourceName())
+		if err != nil {
+			err = errors.Wrapf(err, "failed to get existing instances")
+			s.Scope.UpdatePutStatus(infrav1.BootstrapSucceededCondition, serviceName, err)
+			return err
 		}
-	}()
-
-	if future == nil {
-		fetchedVMSS, err = s.getVirtualMachineScaleSet(ctx, scaleSetSpec.Name)
-	} else {
-		fetchedVMSS, err = s.getVirtualMachineScaleSetIfDone(ctx, future)
+	} else if !azure.ResourceNotFound(err) {
+		return errors.Wrapf(err, "failed to get existing VMSS")
 	}
 
-	switch {
-	case err != nil && !azure.ResourceNotFound(err):
-		// There was an error and it was not an HTTP 404 not found. This is either a transient error, like long running operation not done, or an Azure service error.
-		return errors.Wrapf(err, "failed to get VMSS %s", scaleSetSpec.Name)
-	case err != nil && azure.ResourceNotFound(err):
-		// HTTP(404) resource was not found, so we need to create it with a PUT
-		future, err = s.createVMSS(ctx)
-		if err != nil {
-			return errors.Wrap(err, "failed to start creating VMSS")
+	result, err := s.CreateOrUpdateResource(ctx, scaleSetSpec, serviceName)
+	s.Scope.UpdatePutStatus(infrav1.BootstrapSucceededCondition, serviceName, err)
+
+	if err == nil && result != nil {
+		vmss, ok := result.(compute.VirtualMachineScaleSet)
+		if !ok {
+			return errors.Errorf("%T is not a compute.VirtualMachineScaleSet", result)
 		}
-	case err == nil:
-		// HTTP(200)
-		// VMSS already exists and may have changes; update it with a PATCH
-		// we do this to avoid overwriting fields in networkProfile modified by cloud-provider
-		future, err = s.patchVMSSIfNeeded(ctx, fetchedVMSS)
-		if err != nil {
-			return errors.Wrap(err, "failed to start updating VMSS")
+
+		fetchedVMSS := converters.SDKToVMSS(vmss, scaleSetSpec.VMSSInstances)
+		if err := s.Scope.ReconcileReplicas(ctx, &fetchedVMSS); err != nil {
+			return errors.Wrap(err, "unable to reconcile replicas")
 		}
-	}
 
-	// Try to get the VMSS to update status if we have created a long running operation. If the VMSS is still in a long
-	// running operation, getVirtualMachineScaleSetIfDone will return an azure.WithTransientError and requeue.
-	if future != nil {
-		fetchedVMSS, err = s.getVirtualMachineScaleSetIfDone(ctx, future)
+		// Transform the VMSS resource representation to conform to the cloud-provider-azure representation
+		providerID, err := azprovider.ConvertResourceGroupNameToLower(azureutil.ProviderIDPrefix + fetchedVMSS.ID)
 		if err != nil {
-			return errors.Wrapf(err, "failed to get VMSS %s after create or update", scaleSetSpec.Name)
+			return errors.Wrapf(err, "failed to parse VMSS ID %s", fetchedVMSS.ID)
```

Comment:
Thanks for noticing and fixing this–there's nothing worse than a misleading error message.